paddle/framework/operator.h

```diff
@@ -281,51 +281,47 @@ class OpKernel {
 
 class OperatorWithKernel : public OperatorBase {
  public:
-  struct OpKernelKey {
-    platform::Place place_;
-
-    OpKernelKey() = default;
-    explicit OpKernelKey(const platform::DeviceContext& dev_ctx) {
-      place_ = dev_ctx.GetPlace();
-    }
-
+  struct OpKernelKey : public platform::Place {
+    explicit OpKernelKey(platform::Place p) : platform::Place(p) {}
     bool operator==(const OpKernelKey& o) const {
-      return platform::places_are_same_class(place_, o.place_);
+      return platform::places_are_same_class(*this, o);
     }
   };
 
-  struct OpKernelHash {
-    std::hash<bool> hash_;
+  struct OpKernelHasher {
     size_t operator()(const OpKernelKey& key) const {
-      return hash_(platform::is_gpu_place(key.place_));
+      return static_cast<size_t>(platform::is_gpu_place(key));
     }
   };
 
-  using OpKernelMap =
-      std::unordered_map<OpKernelKey, std::unique_ptr<OpKernel>, OpKernelHash>;
+  typedef std::unordered_map<
+      std::string /*op_type*/,
+      std::unordered_map<OpKernelKey, std::unique_ptr<OpKernel>,
+                         OpKernelHasher>>
+      OpDeviceKernelMap;
 
   void InferShape(const Scope& scope) const override {
     InferShape(InferShapeContext(*this, scope));
   }
 
   void Run(const Scope& scope,
            const platform::DeviceContext& dev_ctx) const final {
-    auto& opKernel = AllOpKernels().at(type_).at(OpKernelKey(dev_ctx));
-    opKernel->Compute(ExecutionContext(*this, scope, &dev_ctx));
+    AllOpKernels()[type_][OpKernelKey(dev_ctx.GetPlace())]->Compute(
```

Comment:
when use `at`, it will throw error if key is not in the map, like:
```shell
terminating with uncaught exception of type std::out_of_range: map::at:  key not found
```

but if we use `[]`, it will just run even if the key not exist, and cause segment fault at some other place.